software/speller/data_collection_platform/dcp/api.py

```diff
@@ -38,38 +39,128 @@ def index():
+        subprocess_dict['frequency'] = float(data.get('frequency'))
+        subprocess_dict['phase'] = float(data.get('phase'))
+    except:
+        return {'Invalid Request': 'Could not convert post form data. Make sure the data is the correct type. (string for character, and float for phase and frequency)'}, 400
+    
+    subprocess_dict['state'] = 'collect'
+    return {'Success': 'BCI is collecting'}, 201
+
+@bp.route('/openbci/<int:process_id>/collect/stop', methods=['POST'])
+def openbci_process_collect_stop(process_id:int):
+    if process_id not in shared.bci_processes_states:
+        return {'Invalid Request': 'There is no process with this id, make sure your process id is valid'}, 404
+    subprocess_dict = shared.bci_processes_states[process_id]
+    subprocess_dict['state'] = 'ready'
+
+    # write to database 
+    if not write_stream_data(subprocess_dict):
+        return {'result': "Did not write any new information to database, stopping collection"}, 202
+    # clear the subprocess_dict
+    subprocess_dict['character'] = None
+    subprocess_dict['frequency'] = None
+    subprocess_dict['phase'] = None
+    subprocess_dict['q'] = deque()
+
+    return {'Success': 'BCI has stopped collecting, and the queue has been writen to the db'}, 201
 
 @bp.route("/openbci/<int:process_id>/stop", methods=['POST'])
 def openbci_stop(process_id: int):
-    # TODO: Might terminate the wrong process (!!).
-    # We need some kind of process manager, and call
-    # mp.Process.terminate() and mp.Process.kill() directly.
-    # This would also allow us to use mp.Process.is_alive().
-    shared.bci_processes_states[process_id] = False
-    os.kill(process_id, signal.SIGTERM)
-    return {}, 200
+    if process_id not in shared.bci_processes_states:
+        return {'Invalid Request': 'There is no process with this id, make sure your process id is valid'}, 404
+    subprocess_dict = shared.bci_processes_states[process_id]
+    if subprocess_dict['state'] == 'collect':
+        subprocess_dict['state'] = 'stop'
+        return {"error": f"stopped bci process while it was collecting, data was not written for character {subprocess_dict['character']}"}, 400
+    
+    subprocess_dict['state'] = 'stop'
+    if subprocess_dict['q']:
+        return {'warning': f"stopped bci process, however the queue for bci data was not empty, data for character {subprocess_dict['character']} might be incomplete"}, 400
+    shared.bci_processes_states.pop(process_id, None)
+    return {'success': f"ended subprocess with id {process_id}"}, 200
+
+def write_stream_data(subprocess_dict):
+    # ensure any remaining connections are flushed to avoid racing conditions
+    db.engine.dispose()
+    order = 1
+    queue = subprocess_dict['q']
+    collected_data = []
+    while queue:
+        stream_data = queue.popleft()
+        for row in stream_data:
+            if len(row) < 8:
+                raise Exception("BCI Data is not the right format")
```

Comment:
we are raising an exception but we are never catching it. However, I think we can safely remove this check, because sometimes the data team will not be using all 8 channels of the BCI device, therefore in those scenarios, I believe there will still be 8 floats in the array (@michellewang please confirm), and the channels that are not used should have values of 0.